src/runtime/disco/worker.h

```diff
@@ -0,0 +1,184 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*!
+ * \file worker.h
+ * \brief This file defines a worker in Disco. A worker can be launched in a separate thread,
+ * process or node as long as the channel supports bi-directional communication in-between the
+ * worker and the controler.
+ */
+#ifndef TVM_RUNTIME_DISCO_WORKER_H_
+#define TVM_RUNTIME_DISCO_WORKER_H_
+
+#include <tvm/runtime/packed_func.h>
+
+#include <atomic>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <vector>
+
+namespace tvm {
+namespace runtime {
+
+class DiscoWorker;
+
+/*!
+ * \brief Controler-to-worker commands.
+ * Each command is broadcasted to all workers.
+ *
+ * Worker-0 is a special worker that is always co-located with the controller process,
+ * and thus could conveniently exchange data with the controler in memory.
+ */
+enum class DiscoAction : int32_t {
+  kShutDown = 0,
+  kKillReg = 1,
+  kGetGlobalFunc = 2,
+  kCallPacked = 3,
+  kSyncWorker = 4,
+  kCopyFromWorker0 = 5,
+  kCopyToWorker0 = 6,
+};
+
+inline std::string DiscoAction2String(DiscoAction action) {
+  switch (action) {
+    case DiscoAction::kShutDown:
+      return "kShutDown";
+    case DiscoAction::kKillReg:
+      return "kKillReg";
+    case DiscoAction::kGetGlobalFunc:
+      return "kGetGlobalFunc";
+    case DiscoAction::kCallPacked:
+      return "kCallPacked";
+    case DiscoAction::kSyncWorker:
+      return "kSyncWorker";
+    case DiscoAction::kCopyFromWorker0:
+      return "kCopyFromWorker0";
+    case DiscoAction::kCopyToWorker0:
+      return "kCopyToWorker0";
+  }
+  LOG(FATAL) << "ValueError: Unknown DiscoAction: " << static_cast<int>(action);
+}
+
+/*! \brief Possible kinds of allreduce or reduce operations. */
+enum class ReduceKind : int32_t {
+  kSum = 0,
+  kProd = 1,
+  kMin = 2,
+  kMax = 3,
+  kAvg = 4,
+};
+
+inline std::string ReduceKind2String(ReduceKind kind) {
+  switch (kind) {
+    case ReduceKind::kSum:
+      return "kSum";
+    case ReduceKind::kProd:
+      return "kProd";
+    case ReduceKind::kMin:
+      return "kMin";
+    case ReduceKind::kMax:
+      return "kMax";
+    case ReduceKind::kAvg:
+      return "kAvg";
+  }
+  LOG(FATAL) << "ValueError: Unknown ReduceKind: " << static_cast<int>(kind);
+}
+
+/*! \brief A bi-directional channel for controler-worker communication. */
```

Comment:
```suggestion
/*! \brief A bi-directional command channel for controler-worker communication. */
```
To emphasize this is not for transferring ndarray data.

Reply:
Maybe stating 

```
A bi-directional channel for controler-worker communication. This channel is primarily used to transfer control messages and not data.
```